### ISSUE TITLE
FMRC HDF5 NullPointerException (DataBTree?)

### ISSUE THREAD

**Issue by rjdave** [Unidata/thredds#527]
I believe this issue is very closely related to #518

I am experiencing the identical top 11 lines of the error described in #518 (besides path2=...) with TDS 4.6.5 on an FMRC dataset of CODAR totals. After line 11 the errors start to differ because of datatypes, dimensions and tomcat and java versions. I have text files with the stack traces spaced out to facilitate diffing if that will help. There is no forecast overlap in this case and the files are compressed NetCDF4. We are using FMRC because NcML aggregation can't handle the 90K files associated with this ever growing dataset.

Unfortunately the test server I'm using is not publicly accessible but, if necessary, we could make it accessible temporarily. The same exact dataset served via TDS 4.6.3 is publicly accessible from:

http://tds.marine.rutgers.edu/thredds/catalog/cool/codar/totals/5Mhz_6km_realtime_fmrc/catalog.html

The initial data request that triggers the error is not the same each time but once triggered that particular data request will continue to throw the error until the server is restarted. We had the same problem with 4.6.4 but I don't have the 4.6.4 log files anymore so I don't know if the stack trace of the error was the same. We do not experience this error with 4.6.3. I have also included the stack trace from the Matlab session that triggered the error. 

In this case the request was:
```
2016-04-12T07:36:54.556 -0400 [ 321551348][     304] INFO  - threddsServlet - Remote host: 192.168.30.10 - Request: "GET /thredds/dodsC/cool/codar/totals/5Mhz_6km_realtime_fmrc/Maracoos_5MHz_6km_Totals-FMRC_best.ncd.dods?u_err[5001:1:5500][0:1:184][0:1:154] HTTP/1.1"
```
and the stack trace is:
```
2016-04-12T07:36:54.813 -0400 [ 321551605][     304] ERROR - thredds.server.opendap.OpendapServlet - path2= /cool/codar/totals/5Mhz_6km_realtime_fmrc/Maracoos_5MHz_6km_Totals-FMRC_best.ncd.dods
java.lang.NullPointerException
    at ucar.unidata.io.RandomAccessFile.read_(RandomAccessFile.java:760) ~[cdm-4.6.5.jar:4.6.5]
    at ucar.unidata.io.RandomAccessFile.readBuffer(RandomAccessFile.java:520) ~[cdm-4.6.5.jar:4.6.5]
    at ucar.unidata.io.RandomAccessFile.seek(RandomAccessFile.java:508) ~[cdm-4.6.5.jar:4.6.5]
    at ucar.nc2.iosp.hdf5.DataBTree$Node.<init>(DataBTree.java:170) ~[cdm-4.6.5.jar:4.6.5]
    at ucar.nc2.iosp.hdf5.DataBTree$DataChunkIterator.<init>(DataBTree.java:135) ~[cdm-4.6.5.jar:4.6.5]
    at ucar.nc2.iosp.hdf5.DataBTree.getDataChunkIteratorFilter(DataBTree.java:61) ~[cdm-4.6.5.jar:4.6.5]
    at ucar.nc2.iosp.hdf5.H5tiledLayoutBB.<init>(H5tiledLayoutBB.java:110) ~[cdm-4.6.5.jar:4.6.5]
    at ucar.nc2.iosp.hdf5.H5iosp.readData(H5iosp.java:162) ~[cdm-4.6.5.jar:4.6.5]
    at ucar.nc2.iosp.hdf5.H5iosp.readData(H5iosp.java:141) ~[cdm-4.6.5.jar:4.6.5]
    at ucar.nc2.NetcdfFile.readData(NetcdfFile.java:2009) ~[cdm-4.6.5.jar:4.6.5]
    at ucar.nc2.Variable.reallyRead(Variable.java:874) ~[cdm-4.6.5.jar:4.6.5]
    at ucar.nc2.Variable._read(Variable.java:845) ~[cdm-4.6.5.jar:4.6.5]
    at ucar.nc2.Variable.read(Variable.java:723) ~[cdm-4.6.5.jar:4.6.5]
    at ucar.nc2.dataset.VariableDS.reallyRead(VariableDS.java:523) ~[cdm-4.6.5.jar:4.6.5]
    at ucar.nc2.dataset.VariableDS._read(VariableDS.java:507) ~[cdm-4.6.5.jar:4.6.5]
    at ucar.nc2.dataset.VariableDS._read(VariableDS.java:531) ~[cdm-4.6.5.jar:4.6.5]
    at ucar.nc2.Variable.read(Variable.java:709) ~[cdm-4.6.5.jar:4.6.5]
    at ucar.nc2.ft.fmrc.FmrcDataset.read(FmrcDataset.java:1182) ~[cdm-4.6.5.jar:4.6.5]
    at ucar.nc2.ft.fmrc.FmrcDataset.access$200(FmrcDataset.java:87) ~[cdm-4.6.5.jar:4.6.5]
    at ucar.nc2.ft.fmrc.FmrcDataset$ProxyReader1D.reallyRead(FmrcDataset.java:1086) ~[cdm-4.6.5.jar:4.6.5]
    at ucar.nc2.dataset.VariableDS._read(VariableDS.java:537) ~[cdm-4.6.5.jar:4.6.5]
    at ucar.nc2.Variable.read(Variable.java:709) ~[cdm-4.6.5.jar:4.6.5]
    at ucar.nc2.Variable.read(Variable.java:683) ~[cdm-4.6.5.jar:4.6.5]
    at thredds.server.opendap.NcSDArray.read(NcSDArray.java:115) ~[classes/:4.6.5]
    at thredds.server.opendap.NcSDGrid.read(NcSDGrid.java:73) ~[classes/:4.6.5]
    at opendap.servers.SDGrid.serialize(SDGrid.java:415) ~[opendap-4.6.5.jar:4.6.5]
    at opendap.servers.CEEvaluator.send(CEEvaluator.java:315) ~[opendap-4.6.5.jar:4.6.5]
    at thredds.server.opendap.OpendapServlet.doGetDAP2Data(OpendapServlet.java:531) ~[classes/:4.6.5]
    at thredds.server.opendap.OpendapServlet.doGet(OpendapServlet.java:218) [classes/:4.6.5]
    at javax.servlet.http.HttpServlet.service(HttpServlet.java:622) [servlet-api.jar:?]
    at javax.servlet.http.HttpServlet.service(HttpServlet.java:729) [servlet-api.jar:?]
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:292) [catalina.jar:8.0.32]
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:207) [catalina.jar:8.0.32]
    at thredds.servlet.filter.CookieFilter.doFilter(CookieFilter.java:55) [classes/:4.6.5]
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:240) [catalina.jar:8.0.32]
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:207) [catalina.jar:8.0.32]
    at thredds.servlet.filter.RequestQueryFilter.doFilter(RequestQueryFilter.java:118) [classes/:4.6.5]
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:240) [catalina.jar:8.0.32]
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:207) [catalina.jar:8.0.32]
    at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52) [tomcat-websocket.jar:8.0.32]
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:240) [catalina.jar:8.0.32]
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:207) [catalina.jar:8.0.32]
    at thredds.servlet.filter.RequestCORSFilter.doFilterInternal(RequestCORSFilter.java:49) [classes/:4.6.5]
    at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) [spring-web-4.1.6.RELEASE.jar:4.1.6.RELEASE]
    at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:344) [spring-web-4.1.6.RELEASE.jar:4.1.6.RELEASE]
    at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:261) [spring-web-4.1.6.RELEASE.jar:4.1.6.RELEASE]
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:240) [catalina.jar:8.0.32]
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:207) [catalina.jar:8.0.32]
    at thredds.servlet.filter.RequestPathFilter.doFilter(RequestPathFilter.java:94) [classes/:4.6.5]
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:240) [catalina.jar:8.0.32]
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:207) [catalina.jar:8.0.32]
    at thredds.server.RequestBracketingLogMessageFilter.doFilter(RequestBracketingLogMessageFilter.java:81) [classes/:4.6.5]
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:240) [catalina.jar:8.0.32]
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:207) [catalina.jar:8.0.32]
    at org.apache.logging.log4j.web.Log4jServletFilter.doFilter(Log4jServletFilter.java:71) [log4j-web-2.2.jar:2.2]
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:240) [catalina.jar:8.0.32]
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:207) [catalina.jar:8.0.32]
    at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:212) [catalina.jar:8.0.32]
    at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:106) [catalina.jar:8.0.32]
    at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:502) [catalina.jar:8.0.32]
    at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:141) [catalina.jar:8.0.32]
    at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:79) [catalina.jar:8.0.32]
    at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:616) [catalina.jar:8.0.32]
    at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:88) [catalina.jar:8.0.32]
    at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:522) [catalina.jar:8.0.32]
    at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1095) [tomcat-coyote.jar:8.0.32]
    at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:672) [tomcat-coyote.jar:8.0.32]
    at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1500) [tomcat-coyote.jar:8.0.32]
    at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1456) [tomcat-coyote.jar:8.0.32]
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_74]
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_74]
    at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) [tomcat-util.jar:8.0.32]
    at java.lang.Thread.run(Thread.java:745) [?:1.8.0_74]
```
and the Matlab error was:
```
opendap.dap.DataReadException: Inconsistent array length read: 1165128303 != 1914731274
	at opendap.dap.DVector.deserialize(DVector.java:251)
	at opendap.dap.DGrid.deserialize(DGrid.java:446)
	at opendap.dap.DataDDS.readData(DataDDS.java:117)
	at opendap.dap.DConnect2$DataDDSCommand.process(DConnect2.java:794)
	at opendap.dap.DConnect2.openConnection(DConnect2.java:311)
	at opendap.dap.DConnect2.getData(DConnect2.java:778)
	at opendap.dap.DConnect2.getData(DConnect2.java:1058)
	at ucar.nc2.dods.DODSNetcdfFile.readDataDDSfromServer(DODSNetcdfFile.java:1172)
	at ucar.nc2.dods.DODSNetcdfFile.readData(DODSNetcdfFile.java:1336)
	at ucar.nc2.Variable.reallyRead(Variable.java:847)
	at ucar.nc2.Variable._read(Variable.java:833)
	at ucar.nc2.Variable.read(Variable.java:645)
	at ucar.nc2.Variable.read(Variable.java:591)
opendap.dap.DAP2Exception: opendap.dap.DataReadException: Inconsistent array length read: 1165128303 != 1914731274
	at opendap.dap.DConnect2.openConnection(DConnect2.java:315)
	at opendap.dap.DConnect2.getData(DConnect2.java:778)
	at opendap.dap.DConnect2.getData(DConnect2.java:1058)
	at ucar.nc2.dods.DODSNetcdfFile.readDataDDSfromServer(DODSNetcdfFile.java:1172)
	at ucar.nc2.dods.DODSNetcdfFile.readData(DODSNetcdfFile.java:1336)
	at ucar.nc2.Variable.reallyRead(Variable.java:847)
	at ucar.nc2.Variable._read(Variable.java:833)
	at ucar.nc2.Variable.read(Variable.java:645)
	at ucar.nc2.Variable.read(Variable.java:591)
Caused by: opendap.dap.DataReadException: Inconsistent array length read: 1165128303 != 1914731274
	at opendap.dap.DVector.deserialize(DVector.java:251)
	at opendap.dap.DGrid.deserialize(DGrid.java:446)
	at opendap.dap.DataDDS.readData(DataDDS.java:117)
	at opendap.dap.DConnect2$DataDDSCommand.process(DConnect2.java:794)
	at opendap.dap.DConnect2.openConnection(DConnect2.java:311)
	... 8 more
Error using nc_varget_java>nc_vara_get_java (line 220)
Java exception occurred:
java.io.IOException: opendap.dap.DataReadException: Inconsistent array length read: 1165128303 != 1914731274
	at ucar.nc2.dods.DODSNetcdfFile.readData(DODSNetcdfFile.java:1343)
	at ucar.nc2.Variable.reallyRead(Variable.java:847)
	at ucar.nc2.Variable._read(Variable.java:833)
	at ucar.nc2.Variable.read(Variable.java:645)
	at ucar.nc2.Variable.read(Variable.java:591)

Error in nc_varget_java>nc_vara_get_java (line 220)
values = jvarid.read(start, count);

Error in nc_varget_java (line 102)
            values = nc_vara_get_java(jvarid,theDataTypeString,start,count);

Error in nc_varget (line 49)
		data = nc_varget_java(ncfile,varname,start,count,stride);
```

**Comment by rsignell-usgs** [comment]
@rjdave , did this issue get subsumed elsewhere?

**Comment by rjdave** [comment]
@rsignell-usgs Not that I'm aware of. Maybe it's not being actively looked at since @dblodgett-usgs found a workaround for his similar situation #518.

**Comment by dblodgett-usgs** [comment]
My workaround was eventually to downgrade to 4.5.5 with a war overlay fix for the malformed ddx xml. I really haven't found a 4.6.x release that is production ready.

**Comment by rsignell-usgs** [comment]
> I really haven't found a 4.6.x release that is production ready.

Ouch.  @lesserwhirls, what say ye?

**Comment by lesserwhirls** [comment]
The way caching was done in version v4.6.4 was tweaked, which I think is when we first got reports on the DataBTree issue, but that was change was reverted in v4.6.5, so we thought these would go away. 

@dblodgett-usgs - I thought in [501](https://github.com/Unidata/thredds/issues/501#issuecomment-203068104), you said v4.6.3 was doing ok? Were there other issues with v4.6.3 that make it not production ready?

So, in trying to sort this out:

@rjdave 
v4.6.5 using FMRC - error
v4.6.4 using FMRC - error
v4.6.3 using FMRC - it works

@dblodgett-usgs
v4.6.5 using NcML - error
v4.6.4 using NcML - error
v4.6.3 using NcML - works?

@cwardgar is currently looking into Unidata/thredds#518, as it seems to be more reproducible. Once a fix is found for that issue, then we can see if this intermittent failure is fixed.

**Comment by rjdave** [comment]
the @rjdave section is correct. We continue to run 4.6.3 as our production server.

**Comment by dblodgett-usgs** [comment]
@lesserwhirls #541 is a real problem for us. I'll work up the examples you need on our testing server this weekend.

**Comment by JohnLCaron** [comment]
I think this should be fixed by #518 ?

**Comment by cwardgar** [comment]
@rjdave I believe this was fixed in #518, so I'm closing it. [Here](tds-4.6.7-20160804.233527-4.war) is a snapshot of TDS that contains the fix, if you'd like to try it. Rename the file to "thredds.war" before you deploy into Tomcat.
